svgdigitizer/svgplot.py

```diff
@@ -197,29 +208,163 @@ class SVGPlot:
+            ...   <g>
+            ...     <path d="M -100 0 L 0 0" />
+            ...     <text x="-100" y="0">j2: 1 A</text>
+            ...   </g>
+            ... </svg>'''))
+            >>> plot = SVGPlot(svg)
+            >>> plot.xlabel
+            'E'
+        """
+
+        [axis] = [
+            axis
+            for axis, orientation in self.axis_orientations.items()
+            if orientation == AxisOrientation.HORIZONTAL
+        ]
+        return axis
+
+    @property
+    @cache
+    def ylabel(self):
+        r"""
+        Return the label of the y axis.
+
+        EXAMPLES::
+
+            >>> from svgdigitizer.svg import SVG
+            >>> from io import StringIO
+            >>> svg = SVG(StringIO(r'''
+            ... <svg>
+            ...   <g>
+            ...     <path d="M 0 200 L 0 100" />
+            ...     <text x="0" y="200">E1: 0 cm</text>
+            ...   </g>
+            ...   <g>
+            ...     <path d="M 100 200 L 100 100" />
+            ...     <text x="100" y="200">E2: 1cm</text>
+            ...   </g>
+            ...   <g>
+            ...     <path d="M -100 100 L 0 100" />
+            ...     <text x="-100" y="100">j1: 0</text>
+            ...   </g>
+            ...   <g>
+            ...     <path d="M -100 0 L 0 0" />
+            ...     <text x="-100" y="0">j2: 1 A</text>
+            ...   </g>
+            ... </svg>'''))
+            >>> plot = SVGPlot(svg)
+            >>> plot.ylabel
+            'j'
+        """
+
+        [axis] = [
+            axis
+            for axis, orientation in self.axis_orientations.items()
+            if orientation == AxisOrientation.VERTICAL
+        ]
+        return axis
+
+    @property
+    @cache
+    def axis_orientations(self):
+        r"""
+        Return the orientation for each axis.
+        The current implementation does not work for extrem cases where a plot is tilted by 45°.
+
+        EXAMPLES::
+
+            >>> from svgdigitizer.svg import SVG
+            >>> from io import StringIO
+            >>> svg = SVG(StringIO(r'''
+            ... <svg>
+            ...   <g>
+            ...     <path d="M 0 200 L 0 100" />
+            ...     <text x="0" y="200">E1: 0 cm</text>
+            ...   </g>
+            ...   <g>
+            ...     <path d="M 100 200 L 100 100" />
+            ...     <text x="100" y="200">E2: 1cm</text>
+            ...   </g>
+            ...   <g>
+            ...     <path d="M -100 100 L 0 100" />
+            ...     <text x="-100" y="100">j1: 0</text>
+            ...   </g>
+            ...   <g>
+            ...     <path d="M -100 0 L 0 0" />
+            ...     <text x="-100" y="0">j2: 1 A</text>
+            ...   </g>
+            ... </svg>'''))
+            >>> plot = SVGPlot(svg)
+            >>> plot.axis_orientations
+            {'E': <AxisOrientation.HORIZONTAL: 'horizontal'>, 'j': <AxisOrientation.VERTICAL: 'vertical'>}
+        """
+
+        from numpy import trace
```

Comment:
It's a surprisingly reasonable approach to use the trace here. Since you told me that you are just guessing by looking at a few examples, let me try to explain why this is not a bad idea.

The affine transformation matrix is a 3×3 matrix. Essentially consisting of a 2×2 matrix in the top-left corner that describes a linear transformation and a 2×1 matrix in the top right that describes a translation. The rest of the matrix is most of the time just `0 0 1` in the bottom row.

(Not terribly import here, but your approach obviously fails if the bottom row is different in the two transformation matrices. Probably this does not happen here.)

So the question is: Looking at that 2×2 part, how can we estimate the amount of rotation that is happening?

Let's pretend for a moment that we just have a linear transformation and only look at the 2×2 matrix part and let's pretend that it's just a rotation by the angle θ. In that case, it's actually true that the trace is 1 + 2 cosθ. So, given the trace we can recover the absolute value of the angle. And more importantly, a small trace means rotation by a large angle, a small trace rotation by a large angle, at least when these angles are less that ±90°. (And a similar statement is true for angles close to 180° btw.)

The above explains why your approach works when the transformation is only a rotation.

Now what happens when the axes are differently scaled, i.e., when the transformation is not only a rotation matrix but a rotation matrix times a diagonal matrix. I think the approach still works because the scaling is compatible in the two transformations you compute.

I am not actually sure what happens when there is also shear. It might magically work again because the shear is somehow consistent in the two cases but I am not sure.

I think the safe approach here is to decompose the 2×2 matrix into a (rotation with flip) and a shear matrix. This is called a QR decomposition. Of this QR decomposition we can then compare the rotation angle with your approach.

---

So here's what we should do:

Grab the top left 2×2 matrix that describes the linear transformation of the affine transformation. QR decompose this matrix, this gives us a Q matrix which is orthogonal, i.e., a rotation, a reflection, or a combination of both. If the determinant of Q is 1, it's a rotation and so its trace measures the amount of rotation (the bigger the trace, the smaller the rotation.) If the determinant is -1, it's a reflection or a rotation+reflection. We can ignore the case that it's just a reflection since it's just a reflection and a rotation by 0°. So let's say it's a rotation+reflection. Then we undo the reflection by swapping the rows of the matrix Q and compute its trace and again, a large trace means little rotation, a small trace means big rotation.

For this to work, we do not even need the bottom row of the affine transformations to be the same `0 0 c`. It works even if they differ.

Reply:
> Then we undo the reflection by swapping the rows of the matrix Q and compute its trace and again, a large trace means little rotation, a small trace means big rotation.

Sorry, that's nonsense. Such a swap corresponds to a flip with axis x=y.